### PR TITLE
Use REST spenttxouts for block indexing, plus consistency fixes and RocksDB multi-CF

### PR DESCRIPTION
**Pull request by shesek**
This PR adds a new indexing mode using the Bitcoin Core `spenttxouts` REST endpoint to look up spent prevouts and switches to REST for raw block contents (in both indexing modes).

The existing indexing mode operates in two phases: first populate `txstore`, then populate `history` while looking up spent prevouts in `txstore`. This makes indexing read-heavy and introduces an ordering requirement: before a block can be indexed into `history`, it and all its ancestors must first be populated into `txstore`.

The new indexer fetches each block alongside its spent prevouts directly from the Bitcoin Core binary REST API using the `block` and `spenttxouts` endpoints, then writes the complete `txstore`+`history` DB rows for each block. This allows blocks to be indexed in any order with full parallelism, as a write-only workload with no local RocksDB reads.

The new indexer is now the default operating mode for Bitcoin, but the old indexer is selected automatically when `spenttxouts` is not supported: for Elements/Liquid, where it is the only mode, or for Bitcoin Core prior to v30.

Commits can be grouped into:

- **Cleanup**: removes obsolete operating modes that are no longer supported, to reduce maintenance burden when making indexer/schema changes. This includes DB migration, BLK import support and light-mode. (d1804b0a, d7ca30aa, 3104b41e)

- **Groundwork**: prepares shared infrastructure for both indexing modes: switch to ordered prevout consumption, extract reusable code and introduce `HeaderWork`, plus a perf fix. (acfa21e2, 9dcc7adf, 4aea5bb7, 54bdeb8d)

- **Core work**: adds binary REST client support and the `spenttxouts`-based indexer, and tunes RocksDB for increased write throughput. (1ab22fc0, 27448ffb, ab6cf525, 76006eda, 31cf42a9, b4b57ec7)

- **Consistency**: adapts crash recovery, tip publication, stale undo and RocksDB durability for out-of-order block indexing. (18cb9249, af76fa48, cc498cc1, 1303cf91)

### Consistency and crash recovery

Out-of-order block indexing required adapting crash recovery to use a new startup stale cleanup mechanism, where Electrs compares the daemon's best chain with the `D` completion markers and cleans up any stale blocks before normal indexing work. During reorgs, the common ancestor `t` is now persisted before stale `history` deletion so a mid-way crash will recover from a safe tip and complete cleanup.

`Store::open()` also includes a startup fail-safe mechanism from an earlier recovery design (that was discarded), that trims chain visibility down to the fully indexed contiguous prefix. The current design does not depend on this, but I kept it anyway to guard against unexpected/invalid state. This could alternatively panic instead of trim.

Stale undo now uses local RocksDB data instead of the current bitcoind. This makes recovery safe in multi-backend deployments, where Electrs might reconnect to a different daemon that does not have the stale block.

The PR also adds an AI-assisted document describing the safety model, enumerating potential crash-recovery failure cases, and explaining how they are handled. This helped validate the implementation and identified the two (pre-existing) failure cases fixed by af76fa48 and cc498cc1. I wasn't sure whether to include this in the final PR, but ended up keeping it as reference for humans/AIs making consistency/safety-sensitive changes.

### RocksDB Column Families

The index now uses a single RocksDB database with four column families: `default`, `txstore`, `history` and `cache`.

This was necessary to resolve a consistency/durability issue with the prior multi-DB design, and also desirable regardless.

### Potential Followups

- If we could get Elements to add `spenttxouts` support (a [simple](https://github.com/bitcoin/bitcoin/pull/32540) `+109` patch to rest.cpp plus docs/tests), we could consider supporting recent Core/Elements versions only and remove the old indexer mode entirely.
- We could boost indexer performance even more using @RCasatta's [`bitcoin_slices`](https://github.com/RCasatta/bitcoin_slices) crate for allocation-free block parsing. I initially thought to include this alongside the new indexer, but it's big enough of a change to warrant its own PR. We could take inspiration from [`bindex`](https://github.com/romanz/bindex-rs) which uses `bitcoin_slices`.

### Backward Compatibility

This PR breaks DB compatibility and requires a complete reindex.

### CODE REVIEW

**Comment by shesek** [comment]
Shoutout and a big thanks to @romanz that [made this possible](https://github.com/bitcoin/bitcoin/pull/32540)! 🙌

File: src/config.rs [hunk 1]
```diff
@@ -176,16 +171,6 @@ impl Config {
...
-            .arg(
-                Arg::with_name("jsonrpc_import")
-                    .long("jsonrpc-import")
-                    .help("Use JSONRPC instead of directly importing blk*.dat files. Useful for remote full node or low memory system"),
-            )
-            .arg(
-                Arg::with_name("light_mode")
-                    .long("lightmode")
-                    .help("Enable light mode for reduced storage")
-            )
```
Comment (EddieHouston) [inline src/config.rs:213]: As we have been encouraging people to use jsonrpc-import... perhaps when we remove existing startup arguments, we could add some kind of deprecation with error or warning for a while so that when someone still has these set they don't get a panic exit on startup (or crash looping container).

File: src/daemon.rs [hunk 2]
```diff
@@ -370,13 +367,20 @@ impl Daemon {
...
+                    .thread_name(|i| format!("daemon-requests-{}", i))
                     .build()
                     .unwrap(),
             ),
+            rest_agent: ureq::Agent::new_with_config(
+                ureq::Agent::config_builder()
+                    .max_idle_connections(daemon_parallelism * 2)
+                    .max_idle_connections_per_host(daemon_parallelism * 2)
+                    .build(),
+            ),
```
Comment (EddieHouston) [inline src/daemon.rs]: This builds the ureq::Agent without setting any timeout, and ureq defaults global/connect/recv_response/recv_body to None. 

The old JSON-RPC TcpStream path uses explicit read/write/connect timeouts (see tcp_connect above ~line 195). 

Could we add some kind of REST timeouts (maybe also use daemon timeouts if it makes sense)?

File: src/daemon.rs [hunk 3]
```diff
@@ -847,4 +855,70 @@ impl Daemon {
...
+    /// Requires bitcoind started with -rest=1.
+    #[cfg(not(feature = "liquid"))]
+    pub fn get_spent_txouts(&self, blockhash: &BlockHash) -> Result<Vec<Vec<bitcoin::TxOut>>> {
+        let url = format!("http://{}/rest/spenttxouts/{}.bin", self.addr, blockhash);
+        let mut response = self.rest_agent.get(&url).call().map_err(|e| {
+            ErrorKind::Connection(format!(
+                "REST spenttxouts failed for {} (is bitcoind running with -rest=1?): {}",
+                blockhash, e
+            ))
+        })?;
```
Comment (EddieHouston) [inline src/daemon.rs]: Should we have retries here like above in pub getblock() (above starting line 581)?

File: src/new_index/schema.rs [hunk 4]
```diff
@@ -334,18 +366,120 @@ impl Indexer {
...
+        new_headers: &[HeaderEntry],
+        chain_tip_height: usize,
+    ) -> Result<()> {
+        let to_process = self.headers_to_process(new_headers);
+
+        if to_process.is_empty() {
+            return Ok(());
+        }
+
+        #[cfg(not(feature = "liquid"))]
```
Comment (EddieHouston) [inline src/new_index/schema.rs:503]: `--no-spenttxouts` is still a supported Bitcoin mode, but the non-Liquid test harness now always sets `use_spenttxouts: true` (`tests/common.rs` line 125). That leaves this refactored legacy Bitcoin path untested; Liquid exercises a similar path, but under different chain types/cfgs.

  Could we add at least one Bitcoin integration test or TestRunner variant with `use_spenttxouts: false`, covering sync plus a spend/history/utxo query?

File: src/config.rs
Comment (EddieHouston) [inline src/config.rs:530]: I'm wondering if we should add a `log.modules(["electrs", "electrs_macros", etc]);` here... 
Reason is that when in trace mode... the new spenttxouts calls route through ureq-proto which hex-dumps all of the `/rest/block/*.bin` and `/rest/spenttxouts/*.bin` responses... and I was piping the logs to a file and ran out of disk space when they got to 1.3 TB 😅   Might be nice to scope the verbosity to just modules we "care about".

File: src/daemon.rs [hunk 5]
```diff
@@ -602,52 +579,83 @@ impl Daemon {
...
-        self.request("getblock", json!([blockhash, verbose]))
-    }
-
-    #[trace]
-    pub fn getblocks(&self, blockhashes: &[BlockHash]) -> Result<Vec<Block>> {
-        let params_list: Vec<Value> = blockhashes
-            .iter()
-            .map(|hash| json!([hash, /*verbose=*/ false]))
-            .collect();
+        let url = format!("http://{}/rest/block/{}.bin", self.addr, blockhash);
```
Comment (EddieHouston) [inline src/daemon.rs]: Codex-discovered review note (blocking): `new-index` now supports `--daemon-rpc-fallback-addr`, but REST requests here always use the fixed primary `addr`. If JSON-RPC fails over, block and `spenttxouts` fetching can continue targeting the unavailable primary.

During the rebase, could we make REST backend selection consistent with daemon failover—ideally trying the active RPC backend first and the alternate backend for retryable failures? It would be good to cover this with a test where the primary becomes unavailable during indexing.

File: src/new_index/schema.rs [hunk 6]
```diff
@@ -415,61 +550,119 @@ impl Indexer {
...
-
-        if let FetchFrom::BlkFiles = self.from {
-            self.from = FetchFrom::Bitcoind;
+        if work.need_history {
+            // Need to flatten() because Bitcoin Core returns spenttxouts as a per-tx vector of
+            // prevouts, while the indexer expects a per-block prevouts vector.
+            let previous_txos: Vec<TxOut> = daemon
+                .get_spent_txouts(block_entry.entry.hash())?
+                .into_iter()
+                .flatten()
```
Comment (EddieHouston) [inline src/new_index/schema.rs:657]: Codex-discovered review note (blocking): Flattening the response loses the per-transaction boundaries before they are validated. `index_block()` verifies the total number of consumed prevouts, but a malformed response with the correct total and incorrect per-transaction counts could associate prevouts with the wrong inputs.

Could we first verify that (1) the response contains `block.txdata.len() - 1` entries and (2) each entry has exactly as many prevouts as the corresponding non-coinbase transaction has inputs? Then flattening it for the existing indexing interface is safe.

File: src/config.rs [hunk 7]
```diff
@@ -510,6 +495,8 @@ impl Config {
             db_write_buffer_size_mb: value_t_or_exit!(m, "db_write_buffer_size_mb", usize),
             initial_sync_batch_size: value_t_or_exit!(m, "initial_sync_batch_size", usize),
             db_cache_index_filter_blocks: m.is_present("cache_index_filter_blocks"),
+            #[cfg(not(feature = "liquid"))]
+            use_spenttxouts: !m.is_present("no_spenttxouts"),
```
Comment (EddieHouston) [inline src/config.rs]: Codex-discovered review note (recommended): `spenttxouts` is now the default, but startup still accepts older Bitcoin Core versions and only discovers missing REST support once indexing begins. Could we perform an early capability check and return an actionable error such as: “spenttxouts indexing requires Bitcoin Core 30+ with `-rest=1`; use `--no-spenttxouts` for older Core”?

This would distinguish an unsupported Core version, disabled REST, and a transient connection failure.

**Comment by EddieHouston** [comment]
Codex-discovered review note (blocking): The consistency document is valuable, but this PR does not appear to add automated tests for the new completion-marker and crash-recovery behavior. Given that out-of-order indexing is the reason these mechanisms changed, I think we should turn the principal recovery cases into tests before merging.

At minimum:

1. txstore complete, history incomplete;
2. later blocks complete while an earlier block is incomplete;
3. completed work exists beyond persisted `t`;
4. restart during stale/reorg cleanup.

Each test should reopen the database and verify both the visible tip and history/UTXO results, rather than only checking marker rows.

File: src/config.rs (hunk 1 above)
Comment (shesek) [reply src/config.rs:213]: Makes sense. I added `--jsonrpc-import` back as a no-op with a deprecation warning in aee06c30.

File: src/daemon.rs (hunk 2 above)
Comment (shesek) [reply src/daemon.rs]: Yes, I configured `ureq` timeouts using the same constants in dd067f72.

File: src/new_index/schema.rs (hunk 4 above)
Comment (shesek) [reply src/new_index/schema.rs:503]: My thinking was that Liquid exercises the relevant code paths for the legacy indexer, and that it could be sufficient since the Bitcoin/Liquid-specific differences all live in the lower-level helpers shared by both indexing modes.

But I agree it's probably not good enough. It seems best to also test this against an older Bitcoin Core release that didn't support `spenttxouts` at all, which is what I implemented in 8958426a.

File: src/config.rs
Comment (shesek) [reply src/config.rs:530]: Huh yeah that's no good >.< Should definitely be addressed.

Rather than hiding dependency logging entirely, it could be useful to keep `warn`-level logs for dependencies (plus potentially `info`/`debug` too in higher `-v` levels, perhaps with a new `-vvv` mode) while hiding `trace`-level logs which cause the log size blow up.

However it seems that `stderrlog` doesn't support setting a per-module logging level, so we have to settle for either hiding dependency logging entirely or showing it at the global logging level.

There's a compromise we can make: have no `-v` (`warn+info`) and one `-v` (`debug`) enable logs for all modules, make `-vv` (`trace`) set `log.modules(["electrs"])`, and treat `-vvv` as enabling `trace` for all modules.

This would result in quite a surprising behavior, though: no `-v` and one `-v` would show dependency logging, while increasing to `-vv` suddenly hides it. But still, it seems better than always hiding it?

Another alternative we can consider is switching from `stderrlog` to [`env_logger`](https://docs.rs/env_logger/), which does support per-module verbosity levels.

File: src/daemon.rs (hunk 5 above)
Comment (shesek) [reply src/daemon.rs]: Implemented in 90a62fac.

I considered a few approaches for this, and ended up letting `Connection` be the source of truth for backend selection and failover. The REST client reuses the active `Connection` address and triggers the `Connection` reconnection mechanism on transport errors.

Recycling is implemented through a new `RestAgent` wrapper. This also required putting the agent behind an `Arc<Mutex<T>>` for interior mutability and synchronization, but the cost should be negligible since the lock is only held for the max-age check and cheap Arc cloning in the typical case (or the `ureq::Agent` re-initialization when recycling).

File: src/new_index/schema.rs (hunk 6 above)
Comment (shesek) [reply src/new_index/schema.rs:657]: This codebase generally assumes bitcoind returns valid consistent replies, there are quite a few other places where we could assert more defensively for malformed replies but currently don't.

I'm not saying we necessarily shouldn't validate in this particular case, but not sure where the line is crossed. wdyt?

File: src/config.rs (hunk 7 above)
Comment (shesek) [reply src/config.rs]: I implemented a variant of your suggestion in e93d17dd: it errors if REST is disabled, but checks for `spenttxouts` support to automatically select the indexing mode, instead of the `--no-spenttxouts` CLI option. When it's unsupported, it logs a warning suggesting to upgrade Bitcoin Core.

Do you think that's better, or would you prefer the explicit `--no-spenttxouts` and an error when the setting is incompatible with the daemon?

**Comment by shesek** [comment]
> The consistency document is valuable, but this PR does not appear to add automated tests [..]

The reason I added the consistency document is that testing the full indexer behavior through the high-level `Indexer` interface can get quite tricky. Testing some edge-case invariants requires mocking the bitcoind side, while others require crashing the indexer midway through and validating electrs' state across restarts.

What can be tested more easily is the underlying `Store` interface, which doesn't fully cover the indexer behavior but captures some aspects of it. I actually did have the `visibility_stops_at_first_one_sided_completion` test initially, but it felt somewhat superficial and I ended up not including it. But thinking this through again, I agree that it's useful and added it back in b5ef6e61 along with two other tip-handling tests, which together cover your "at minimum" cases 1 to 3.

> 4. restart during stale/reorg cleanup.

Testing case 4 gets into trickier territory. I'm not sure whether adding it or the other higher-level tests is worth the complexity at this stage, but I'm open to discussing this. One thing that makes this more appealing is that the same testing infrastructure we would need here (bitcoind mocking and cross-restart tests) would also be valuable for testing edge-case mempool behaviors. I would still leave it for a followup PR though.

> verify both the visible tip and history/UTXO results

We already have other tests tying the visible tip to history/UTXO results. I opted to keep the `Store` recovery tests simpler and only check the visible tip and completion markers. Do you think that's enough, or should they be extended to cover history/UTXO too?

**Comment by shesek** [comment]
Rebased and addressed review comments. I kept the fixes as separate commits to aid review, to be squashed later.

**Comment by EddieHouston** [comment]
> Rebased and addressed review comments. I kept the fixes as separate commits to aid review, to be squashed later.

Thank you, will check them out.

One thing I notice is that `cargo install --path .` is now failing with:
 ```
 Installing electrs v0.4.1 (/Users/ehouston/freeelectrslogs/electrs)
    Updating git repository `https://github.com/Blockstream/rust-electrum-client`
    Updating git repository `https://github.com/shesek/electrumd`
error: failed to compile `electrs v0.4.1 (/Users/ehouston/freeelectrslogs/electrs)`, intermediate artifacts can be found at `/Users/ehouston/freeelectrslogs/electrs/target`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.

Caused by:
  package `electrs v0.4.1 (/Users/ehouston/freeelectrslogs/electrs)` does not have a dependency named `corepc-node`                                                                                                   
```

Looks like the cause is 8958426 ("test(index): Test legacy indexer with Bitcoin Core 29"), which moved the Core version selector out of the dev-dependency and into a default-on feature:

```
[features]
default = ["test-core-30"]
test-core-30 = ["corepc-node/30_2"]

[dev-dependencies]
corepc-node = { version = "0.12", features = ["download"] }   # was ["download", "30_2"]
```

and cargo install strips dev-dependencies, so once test-core-30 activates it references a dependency that isn't in the install-time manifest. cargo build and cargo test keep dev-deps, which is why CI here is green... only cargo install breaks.

Reproduced on **2938ada**

File: src/daemon.rs (hunk 3 above)
Comment (shesek) [reply src/daemon.rs]: `get_spent_txouts()` now retries through the new `rest_get()` implemented in 90a62fac.

However it is not quite like `getblock()`, which also retries on explicit `not fully downloaded` errors which can occur transiently while new blocks are downloaded. `get_spent_txouts()` is only ever called after `getblock()` completed successfully so it cannot be in that state by that point, but it still retries on transport failures.

**Comment by shesek** [comment]
> cargo install strips dev-dependencies, so once test-core-30 activates it references a dependency that isn't in the install-time manifest

Addressed by eab4df7e1d1a922afee3ab5cc21d6a016cdf06d4.

File: src/daemon.rs [hunk 8]
```diff
@@ -837,11 +874,105 @@ impl Daemon {
...
+        let mut reader = BufReader::new(response.body_mut().as_reader());
+        let block = Block::consensus_decode(&mut reader).chain_err(|| "failed to parse block")?;
+        ensure!(
+            reader
+                .fill_buf()
+                .chain_err(|| "failed to end stream")?
+                .is_empty(),
+            "block response has trailing data"
+        );
+        ensure!(
```
Comment (shesek) [inline src/daemon.rs:1178]: The block hash verification was carried over from the old code, but I think we could consider dropping it. ([related comment](https://github.com/Blockstream/electrs/pull/217#discussion_r3668634068))

**Comment by sammy007** [comment]
Worth bumping rocksdb to 0.25 as well

File: src/new_index/schema.rs (hunk 6 above)
Comment (EddieHouston) [reply src/new_index/schema.rs:657]: Good point. If we assume that bitcoind returns valid, internally consistent responses, then this doesn’t need to be a blocking validation requirement at all. The existing total-consumption check is sufficient to catch integration mistakes that change the overall number of prevouts. 👍

File: src/daemon.rs (hunk 8 above)
Comment (EddieHouston) [reply src/daemon.rs:1178]: Nor critical.. but I could see that this could be removed for same reasons in related comment...

**Comment by EddieHouston** [comment]
> > The consistency document is valuable, but this PR does not appear to add automated tests [..]
> 
> The reason I added the consistency document is that testing the full indexer behavior through the high-level `Indexer` interface can get quite tricky. Testing some edge-case invariants requires mocking the bitcoind side, while others require crashing the indexer midway through and validating electrs' state across restarts.
> 
> What can be tested more easily is the underlying `Store` interface, which doesn't fully cover the indexer behavior but captures some aspects of it. I actually did have the `visibility_stops_at_first_one_sided_completion` test initially, but it felt somewhat superficial and I ended up not including it. But thinking this through again, I agree that it's useful and added it back in [b5ef6e6](https://github.com/Blockstream/electrs/commit/b5ef6e6119501093095c620e154d9b224d4674e7) along with two other tip-handling tests, which together cover your "at minimum" cases 1 to 3.
> 
> > 4. restart during stale/reorg cleanup.
> 
> Testing case 4 gets into trickier territory. I'm not sure whether adding it or the other higher-level tests is worth the complexity at this stage, but I'm open to discussing this. One thing that makes this more appealing is that the same testing infrastructure we would need here (bitcoind mocking and cross-restart tests) would also be valuable for testing edge-case mempool behaviors. I would still leave it for a followup PR though.
> 
> > verify both the visible tip and history/UTXO results
> 
> We already have other tests tying the visible tip to history/UTXO results. I opted to keep the `Store` recovery tests simpler and only check the visible tip and completion markers. Do you think that's enough, or should they be extended to cover history/UTXO too?
 
Thanks. I’m comfortable accepting the current Store tests for cases 1–3.
  
We’re also adding black-box restart/reorg coverage in our internal integration harness, including verification of the visible tip, history, balance, transaction status, and UTXOs after reopening electrs.

I’m okay leaving the more specialized “restart during stale cleanup” case for future work. With that, I no longer consider this comment blocking.

File: src/config.rs
Comment (EddieHouston) [reply src/config.rs:530]: We should probably go with the simplest predictable behavior here... normal CLI verbosity, including the highest commonly used level, should not enable binary REST-body dumps from dependencies.
 
We’re adding an integration-harness regression check that runs electrs at maximum verbosity, indexes a bounded workload, and verifies that ureq/ureq-proto does not dump REST response bodies or produce excessive logs.
 
Could you implement the logging-side fix in this PR? I don’t have a strong preference between changing logger implementations and explicitly filtering dependency trace output, as long as the resulting verbosity behavior is unsurprising.

Comment (shesek) [reply src/config.rs:530]: The `ureq`/`ureq-proto` split actually makes this easier than I thought - `ureq` emits the more useful higher-level logging messages while `ureq-proto` only emits lower-level debug/trace and no warnings/errors, so we can simply omit `ureq-proto`. Implemented in c3f776b68a0eb217624377d7622c535c306dd16b.

**Comment by EddieHouston** [comment]
> Worth bumping rocksdb to 0.25 as well

It seems it's only a few days old.  Checking into it... also created https://github.com/Blockstream/electrs/issues/249

**Comment by EddieHouston** [comment]
> > Worth bumping rocksdb to 0.25 as well
> 
> It seems it's only a few days old. Checking into it... also created #249

@sammy007 Thanks for the heads up on the new 0.25.0.  We plan to do the RocksDB upgrade in a followup PR soon after this lands.

**Comment by EddieHouston** [comment]
Looking more closely at Column Families and see one new concern: 

create_missing_column_families(true) may silently recreate a missing txstore or history CF in an existing database, masking corruption/data loss as recoverable incomplete indexing. 

Could we only create CFs for a fresh DB, and require all expected CFs when reopening an existing one? A reopen test with each required CF missing would validate this.

File: src/new_index/db.rs [hunk 9]
```diff
@@ -464,6 +459,165 @@ impl DB {
...
+        }
+        Some(_) => (),
+    }
+}
+
+fn data_cf_options(config: &Config, shared_cache: &rocksdb::Cache) -> rocksdb::Options {
+    let mut cf_opts = rocksdb::Options::default();
+    cf_opts.set_compaction_style(rocksdb::DBCompactionStyle::Level);
+    cf_opts.set_compression_type(rocksdb::DBCompressionType::Lz4);
+    cf_opts.set_bottommost_compression_type(rocksdb::DBCompressionType::Zstd);
```
Comment (philippem) [inline src/new_index/db.rs:561]: let's consider enabling dictionary here, something like

```
  +    // Zstd dictionary for the bottommost level. Trained per SST file from that
  +    // file's own blocks and stored in its meta block, so it needs no maintenance
  +    // and never goes stale.
  +    if dict_bytes > 0 {
  +        cf_opts.set_bottommost_compression_options(-14, 3, 0, dict_bytes, true);
  +        cf_opts.set_bottommost_zstd_max_train_bytes(dict_bytes * 100, true);
  +    }
```
